make/src/rule/target.rs

```diff
@@ -8,28 +8,99 @@
 //
 
 use core::fmt;
+use crate::special_target::SpecialTarget;
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 /// A target for a rule.
-pub struct Target {
-    name: String,
+pub enum Target {
+    Simple { name: &'static str },
+    Inference {
+        name: &'static str,
+        from: &'static str,
+        to:   &'static str,
+    },
+    Special(SpecialTarget),
 }
 
 impl Target {
     /// Creates a new target with the given name.
     pub fn new(name: impl Into<String>) -> Self {
-        Target { name: name.into() }
+        let name = name.into();
+
+        if let Some(t) = Self::try_parse_special(&name) {
+            return t;
+        }
+
+        if let Some(t) = Self::try_parse_inference(&name) {
+            return t;
+        }
+
+        Target::Simple { name: name.leak() }
+    }
+
+    pub fn name(self) -> &'static str {
+        match self {
+            Target::Simple { name } => name,
+            Target::Inference { name, from, to } => name,
+            Target::Special(target) => match target {
+                SpecialTarget::Default => ".DEFAULT",
+                SpecialTarget::Ignore => ".IGNORE",
+                SpecialTarget::Posix => ".POSIX",
+                SpecialTarget::Precious => ".PRECIOUS",
+                SpecialTarget::SccsGet => ".SCCS_GET",
+                SpecialTarget::Silent => ".SILENT",
+                SpecialTarget::Suffixes => ".SUFFIXES",
+                SpecialTarget::Phony => ".PHONY",
```

Comment:
Isn't `target.as_ref` usable here?

Reply:
Somehow borrowchecker can't infer `'static` lifetime from `as_ref`

Reply:
If so, I think it may be reasonable to add the `as_str` method for cases where `static` lifetime is needed.

But we can leave this one as it is for now.